rad-exe/src/cli/args.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::env;
+
+use structopt::StructOpt;
+
+use librad::profile::ProfileId;
+
+/// `--rad-profile` command line name
+pub const RAD_PROFILE_ARG: &str = "--rad-profile";
+
+/// `--rad-quiet` command line name
+pub const RAD_QUIET_ARG: &str = "--rad-quiet";
+
+/// `--rad-verbose` command line name
+pub const RAD_VERBOSE_ARG: &str = "--rad-verbose";
+
+#[derive(Debug, StructOpt)]
+pub struct Args {
+    /// The profile identifier, if not given then the currently active profile
+    /// is used
+    #[structopt(long)]
+    pub rad_profile: Option<ProfileId>,
+
+    /// No output printed to stdout
+    #[structopt(long)]
+    pub rad_quiet: bool,
+
+    /// Use verbose output
+    #[structopt(long)]
+    pub rad_verbose: bool,
+
+    #[structopt(subcommand)]
+    pub command: Command,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Command {
+    /// This is just a stub, sshhhh
+    // TODO(finto): Fill in core commands
+    Profile,
+    #[structopt(external_subcommand)]
+    External(Vec<String>),
+}
+
+/// If an external subcommand is called, we sanitise the global arguments according to the rules defined in [RFC 698](https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/0698-cli-infrastructure.adoc#global-parameters).
+///
+/// The rules are summarised as:
+///   * The `rad` global _always_ takes precedence, e.g. `rad --rad-profile deaf
+///     xxx --rad-profile beef` will result in `deaf`
+///   * If multiple `rad` globals are given, it is an error, i.e. `rad
+///     --rad-profile deaf --rad-profile beef`
+///   * If multiple globals are given as part of the external command, the last
+///     one is take, e.g. `rad xxx --rad-profile deaf --rad-profile beef` will
+///     result in `beef`.
+///   * Command line arguments take precedence over environment variables, e.g.
+///     `RAD_PROFILE=deaf rad --rad-profile def` will result in `beef`
```

Comment:
Should the rules be extended to also allude to the precedence of commands themselves not only the options/params?

Reply:
Can you provide an example of what you mean?

Reply:
Sorry, the initial comment was highly confusing. What I meant is that the rules state that globals take precedence and I wanted to check if the rules should also state that global/builtin commands take precedence.

Reply:
Ah, well first off, that's not handled by this function, it's handled by the `Command` struct. Secondly, that's stated in the RFC where we state that one shouldn't (in fact can't) override the behaviour of a core command. I say can't because the arg parser will match something like `rad profile` and execute the core command. Whereas it'll fall through if it's `rad xxx`.

Does that make sense?

Reply:
Tippi toppi.